ObsFit: a new cost function implementation for arbitrary observations (sparse, irregular, integrated, etc)

averdy opened the pull request:
## What changes does this PR introduce?
New "ObsFit" package:

An alternative to the pkg/profiles for model-observations comparisons. Given an observational dataset, ObsFit samples the model during the run at the time and location of observations, calculates the cost (sum of weighted misfits), and produces a model-equivalent output file that is directly comparable to the input file. 


## What is the current behaviour? 
The main cost function options are pkg/ecco, which requires gridded datasets at fixed time intervals, and pkg/profiles, which requires a fixed number of depth levels and instantaneous time instances. 


## What is the new behaviour 
With ObsFit, observations do not need to be on a regular grid or a fixed set of depths, making it an efficient tool for sparse and irregular datasets (e.g., altimetry or tomography data). Observations can be made of multiple samples that are averaged or integrated spatially and/or temporally. This implementation will make assimilation of many constraints much more efficient.


## Does this PR introduce a breaking change? 
No


## Other information:
The code has been made available ( https://github.com/averdy/obsfit_mitgcm ) and already has users

It has been implemented in a simple verification experiment ( verification/global_oce_biogeo_bling )

A "README" file includes details on how to run the example, and how to pre-process observations to use with ObsFit

"To do" list:
- ObsFit may not be the best name for this package; alternative suggestions are welcome!
- cost_obsfit.F can be very slow for large datasets. This is where tiled fields are read, combined, and written to a global file. There is probably a more efficient way to do this (help appreciated!)
- I/O routines active_file_* are adapted from pkg/profiles, which has a known problem with tangent linear runs (PR #873)
- Compatibility with LLC grids is in development


## Suggested addition to `tag-index`
New package "ObsFit"

mjlosch commented on model/src/the_main_loop.F:
```diff
@@ -680,6 +687,23 @@ SUBROUTINE THE_MAIN_LOOP( myTime, myIter, myThid )
       ENDIF
 #endif
 
+#ifdef ALLOW_OBSFIT
+      IF (useOBSFIT) THEN
+#ifdef ALLOW_DEBUG
+        IF (debugMode) CALL DEBUG_CALL('obsfit_inloop',myThid)
+#endif
+c--     Accumulate in-situ time averages of theta, salt, and SSH.
+#ifdef ALLOW_AUTODIFF
+C--   Reset the model iteration counter and the model time.
+        myIter = nIter0 + (iloop-1)
+        myTime = startTime + deltaTClock*(iloop-1)
+#endif
+        CALL TIMER_START('OBSFIT_INLOOP      [MAIN_DO_LOOP]', myThid)
+        CALL OBSFIT_INLOOP( myTime, myThid )
+        CALL TIMER_STOP ('OBSFIT_INLOOP      [MAIN_DO_LOOP]', myThid)
+      ENDIF
+#endif
```
I don't understand why `profiles_inloop` and now `obsfit_inloop` needs to be called before the `the_main_loop`. This leads to the extra call before calling the `cost_profiles/obsfit` and generally does not comply with other pkgs. With all other costs or contributions, this is done in `forward_step`, where (to my mind) it belongs. Is there any specific reason to do this differently?

mjlosch replied:
I now understand why this (and `profiles_inloop`) is here: it's the similarity to` cost_averagesfields`.

mjlosch commented on pkg/ecco/ecco_cost_final.F:
```diff
@@ -224,6 +276,18 @@ subroutine ecco_cost_final( myThid )
      &           *objf_profiles_mean(num_var,bi,bj)
       enddo
 #endif
+cav(
+#ifdef ALLOW_OBSFIT
+cav don't add to all tiles
+c      do num_file=1,NFILESMAX_OBS
+c       do num_var=1,NVARMAX
+c          tile_fc(bi,bj) = tile_fc(bi,bj)
+c     &            + mult_obsfit(num_file,num_var)
+c     &            *objf_obsfit(num_file,num_var)
+c       enddo
+c      enddo
+#endif
+cav)
```
is this necessary?

mjlosch commented on pkg/ecco/ecco_cost_final.F:
```diff
@@ -419,6 +514,15 @@ subroutine ecco_cost_final( myThid )
        _GLOBAL_SUM_RL(no_profiles_mean(num_var), myThid )
       enddo
 #endif
+cav(
+#ifdef ALLOW_OBSFIT
+c      do num_file=1,NFILESMAX_OBS
+c       do num_var=1,NVARMAX
+c        _GLOBAL_SUM_RL(f_obsfit(num_file,num_var), myThid )
+c       enddo
+c      enddo
+#endif
+cav)
```
necessary?

mjlosch commented on pkg/ecco/ecco_cost_final.F:
```diff
@@ -379,6 +465,15 @@ subroutine ecco_cost_final( myThid )
        _GLOBAL_SUM_RL(f_profiles_mean(num_var), myThid )
       enddo
 #endif
+cav(
+#ifdef ALLOW_OBSFIT
+c      do num_file_obs=1,NFILESMAX_OBS
+c       do num_var_obs=1,NVARMAX_OBS
+c        _GLOBAL_SUM_RL(f_obsfit(num_file_obs,num_var_obs), myThid )
+c       enddo
+c      enddo
+#endif
+cav)
```
not sure if these commented lines are necessary

averdy replied:
Commented lines have been removed

mjlosch reviewed (changes requested):
Thanks for this addition.
A few comments:
1. I changed the modifications to `verification/global_oce_biogeo_bling` to conform to "secondary" verification experiments: the same `code_ad` but different `input_ad` with specific name `input_ad.obsfit`, see e.g. `global_oce_latlon`. The only addition to `code_ad/packages.conf` was the line `obsfit`, everything else was the same between `code_ad` and `code_ad_obsfit`.
2. In `nput_ad.obsfit/data.pkg`, I added `useGrdchk=.TRUE.,`. There is a `usePROFILES=.TRUE.,` not sure if that's intentional.
3.  after adding a "fake" `results/output_adm.obsfit.txt` (just a copy of `output_adm.txt` for now to trigger testreport to run), `./testreport -adm -ncad -devel -j 6 -t global_oce_biogeo_bling` compiles and gives:
```
Y Y Y Y 16>16<16 16 16 16 16 16 14 16 16 14 13 16 14 16 13 16 13 16 16 16 16 16 16 16 16 pass  global_oce_biogeo_bling  (e=8, w=0, lfd=1, dop=1, sm=0)
Y Y Y N .. .. .. .. .. .. .. .. .. .. .. .. .. .. .. .. .. .. .. N/O   global_oce_biogeo_bling.obsfit
```
more specifically there are taf errors:
```
TAF ERROR    Code of function    nf_inq_varid not seen and no flow information set,
TAF ERROR    assuming all arguments are input
TAF ERROR    Code of function    nf_put_vara_double not seen and no flow information set,
TAF ERROR    assuming all arguments are input
```
and runtime errors:
```
> STOP ABNORMAL END: S/R OBSFIT_INIT_FIXED
> OBSFIT_INIT_FIXED: file obsfit_ssh.nc is not in the pkg/obsfit format
> S/R ALL_PROC_DIE: ending the run
```
that need to be fixed. (It's good that the original verification experiment is not affected!) Once this is fixed, we can have a proper reference output `results/output_adm.obsfit.txt` and start improving details (e.g. your todo-list)
4. The README could be turned into Markdown (README.md) or even better put the content into the documentation
5. If there are issues with LLC or CS grids, it would be a good idea to add a stop somewhere if these grids are used.

averdy commented:
@mjlosch I think I've addressed issues 3-5. Hopefully the verification expt runs now (I can't test it - bad timing).

averdy commented:
Thanks @mjlosch ! I'm getting started on the manual entry. The verification experiment has been updated so that the gradient check is sensitive to the observations. I'll try to check if the nf_open/close are necessary (that part was copied from pkg/profiles). Regarding obsfit_inloop being called before the the_main_loop, maybe @jm-c has thoughts?

IvanaEscobar commented:
Regarding `NF_OPEN/CLOSE`, I found them unnecessary to include in the flow directives. Maybe they were historically required?

gaelforget commented on pkg/obsfit/README.md:
```diff
@@ -0,0 +1,93 @@
+- `sample_lon` (longitude) 
+- `sample_lat` (latitude) 
+- `sample_depth` (depth) 
+
```
not sure why two different name groups -- `obs` vs `sample` -- could this be homogenized to e.g. use `sample` throughout?

averdy replied:
One feature of this package is that it allows observations to be averages or integrals. For example, observations of integrated sound speed along the acoustic ray path. In that case, one specifies multiple locations at which to sample the model. In cost_obsfit, the samples are integrated to make a model-equivalent value, which is compared to the observed value. One "observation" is made of NP "samples". 

In many applications however, NP=1 and "samples" and "observations" are the same. I've been toying with the idea of allowing variables to be called either "sample_lon/lat/depth" or "obs_lon/lat/depth" in the input file, but I don't know if that would make things more confusing or less confusing. Thoughts?

jrscott replied:
I think it is fairly clear as is (also tweaked the main doc on this)

gaelforget commented on pkg/obsfit/README.md:
```diff
@@ -0,0 +1,93 @@
+- I/O `routines active_file_*` are adapted from `pkg/profiles`, which has a known problem with tangent linear runs (PR #873)
+- Compatibility with LLC grids is in development
+
+
```
would be great if this could be added to a new docs chapter dedicated to obsfit

jrscott replied:
@averdy let's discuss the readme in MITgcm dev zoom

gaelforget commented:
> * ObsFit may not be the best name for this package; alternative suggestions are welcome!

couple ideas : 

- sparsedata
- sparseobs
- profiles2 (to highlight the relation to pkg/profiles)
- genprof (as in pkg/ecco's gencost and pkg/ctrl's genctrl)

gaelforget commented:
ping @shoshanareich @mgoldberg10

> > * Compatibility with LLC grids is in development
> 
> * MeshArrays.jl can now be used to get the grid point coefficients and weights for profiles, and should work here too.
> * the original Octave/Matlab package (MITprof) can also be helpful here
> * a python version is being designed (see ECCO HackWeek 2024)
> 
> Would be great to give pointers for all three options in the docs related to this package as well as pkg/profiles, as we have users across packages.
> 
> Ultimately, it would be nice to have this fortran and embedded in MITgcm itself. The main difficulty would be to implement a global knn (or just nearest neighbor) to attach each data point to a tile. The rest could just be ported from MeshArrays.jl to F90, which should not be too difficult I think.
> 
> See "Interpolation Scheme" in [this notebook](https://juliaclimate.github.io/MeshArrays.jl/dev/tutorials/#id_Geography) and
> 
> `(fig1,fig2,fig3)=MeshArrays.plot_examples(:interpolation_demo,Γ)`

averdy commented:
@gaelforget Thanks for the suggestions!

gaelforget commented:
> fig1,fig2,fig3)=MeshArrays.plot_examples(:interpolation_demo,Γ)

To try this out, as mentioned today, you can try via

1.1  [ECCO-Docker](https://github.com/gaelforget/ECCO-Docker#readme)

`docker run -p 8888:8888 gaelforget/ecco-docker` locally, or [binder](https://mybinder.org/v2/gh/gaelforget/ECCO-Docker/HEAD)

1.2 or just [install julia on your laptop](https://julialang.org/downloads/)

2.1 use [Pluto](https://plutojl.org) (see bottom of page) + [notebook url](https://raw.githubusercontent.com/JuliaClimate/MeshArrays.jl/master/examples/geography.jl)

2.2 in the Julia terminal

```
using MeshArrays, CairoMakie
Γ=GridLoad(ID=:LLC90,option="full")
(fig1,fig2,fig3)=MeshArrays.plot_examples(:interpolation_demo,Γ)
```

averdy commented:
This package is now compatible with LLC grids, thanks to @shoshanareich!

shoshanareich commented:
> This package is now compatible with LLC grids, thanks to @shoshanareich!

Yup, let me know if anyone finds any bugs! 

@mgoldberg10 and I are developing [ObsPrep](https://github.com/ECCO-Hackweek/EH24-processors-llc/tree/main/obsprep), a python module for preprocessing datasets to create ObsFit input files for the LLC grid. We created a tutorial on how to use it, which we used to create a test input file you can find on the repo. Let us know if you have any questions!

mjlosch commented:
Just a comment: Much of the NetCDF related discussion applies here, too: `NF_CLOSE`/`NF_OPEN` vs. `NF_SYNC`, (no) error handling, etc.

jm-c commented:
@averdy I merge master-branch in (+ fix a merging conflict) so just be careful, when you get back to this branch, to get the latest (just with 'git pull") before making any new changes (any new commit).

mjlosch reviewed (commented):
This is an incomplete review, because I feel that the PR is not yet ready for review:
1. The AD-verification experiment is triggered by a `output_adm.obsfit.txt` file that appears to be generated by a forward simulation, there's no AD output and no grdchk output. Also the I cannot run the verification experiment, because there is no data.obsfit linked into the run-directory. I can link this manually and then run the executable, but then there's no output from `obsfit_cost` implying that `NFILESMAX_OBS=0` for this experiment. I don't know how to fix that. The gradient check results in O(3.E-03) differences between FD and AD gradient, for the main experiment (with just profiles turned on, but again, I don't see how `obsfit` contributes to the cost function) the differences are of O(1.E-06).
2. There are recomputation warnings for a scalar variable (a mask), that is computed without obvious consequences (I cannot see that)
3. Once this is working, I would not be surprised if the gradient check would reveal that the gradients are entirely off, because the keys in the (ineffective) store directives are very likely wrong.
4. There are many things/issues being carried over from the profiles-package, some of them are currently being fixed in PR #887, and I suggest to either wait until that PR is merged or directly carry over the fixes (and avoid the unfixed issues).

As a way forward, I suggest to first remove the store directives in `obsfit_cost.F` for now, fix the verification experiment setup and generate a reference (with the recomputations), making sure the AD/FD ratios are reasonably small. Then we can fix the recomputations with the correct keys. I find it hard to review code that I don't understand, but once the gradient checks are good, I can't complain anymore.

I strongly suggest to avoid unusual practices such as `do bi = 1,nSx`, ...

I am not sure how far we are with PR #887, but it may be useful to use some of the fixes in that PR rather than use the unfixed `profiles`-pkgs as a model for `obsfit`.

mjlosch commented:
@averdy The experiment does not compile for me, somehow obsfit_cost.F (and possibly more?) is missing. Please check, if you forgot to push local changes.

mjlosch commented:
@averdy still does not compile, now a customised `tamc.h` is missing. I can easily add that, but then we don't know if it's the same as yours, so please add all missing files (and maybe check if it works for you if you after checkout a new fresh copy of the repository without any local modifications).

averdy commented:
@mjlosch I get the same gradients (which are also the gradient values I get with the same observations in "profiles" form), but my "grdchk output (g)" are all exactly 0.

mjlosch commented:
@averdy 
I use the obsfit-branch just as it is; there are no modifications locally, `git status` comes out clean, except for extra directories and files that are irrelevant to this branch. Can you try to run
```
./testreport -adm -ncad -devel -j 6 -t global_oce_biogeo_bling
```
or if you have MPI installed (I need to do this as the executable is getting too big for my laptop/gfortran):
```
./testreport -adm -ncad -devel -j 6 -t global_oce_biogeo_bling -MPI 6 -command 'mpirun -n TR_NPROC ./mitgcmuv_ad'
```
?

Also, how do you generate the gradient values that you get 
> with the same observations in "profiles" form

? Can you provide instructions (or files) so that I can try to reproduce that?

averdy commented:
@mjlosch When I remove the SST cost in data.ecco, the value of (FC1-FC2) is 0 with obsfit. I made another setup with input_ad.profiles; I added a results file so it runs now with the test report. That one gives the same values of ADJ GRAD(FC) as the obsfit one. The branch as modified just now should allow you to repeat that.

mjlosch commented:
@averdy thanks, I can now reproduce the problem:
```
(base) bkli04m056::global_oce_biogeo_bling (obsfit)> grep "grdchk [o ]" tr_run.*/STDOUT.0000
tr_run.obsfit/STDOUT.0000: pkg/grdchk               compiled   and   used ( useGrdchk                = T )
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output h.p:  Id Itile Jtile LAYER   bi   bj   X(Id)           X(Id)+/-EPS
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output h.c:  Id  FC                   FC1                  FC2
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output h.g:  Id     FC1-FC2/(2*EPS)      ADJ GRAD(FC)         1-FDGRD/ADGRD
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   1     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   1  1.5761147921042E+04  1.5761147921042E+04  1.5761147921042E+04
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   1     0.0000000000000E+00 -9.5230937004089E-03  1.0000000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   2     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   2  1.5761147921042E+04  1.5761147921042E+04  1.5761147921042E+04
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   2     0.0000000000000E+00 -7.3835136413574E+01  1.0000000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   3     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   3  1.5761147921042E+04  1.5761147921042E+04  1.5761147921042E+04
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   3     0.0000000000000E+00 -2.8371569824219E+02  1.0000000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   4     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   4  1.5761147921042E+04  1.5761147921042E+04  1.5761147921042E+04
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   4     0.0000000000000E+00 -4.2511917650700E-02  1.0000000000000E+00
tr_run.obsfit/STDOUT.0000:(PID.TID 0000.0001) grdchk  summary  :  RMS of    4 ratios =  1.0000000000000E+00
tr_run.profiles/STDOUT.0000: pkg/grdchk               compiled   and   used ( useGrdchk                = T )
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output h.p:  Id Itile Jtile LAYER   bi   bj   X(Id)           X(Id)+/-EPS
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output h.c:  Id  FC                   FC1                  FC2
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output h.g:  Id     FC1-FC2/(2*EPS)      ADJ GRAD(FC)         1-FDGRD/ADGRD
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   1     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   1  1.5761147921042E+04  1.5761147920946E+04  1.5761147921137E+04
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   1    -9.5231371233240E-03 -9.5230937004089E-03 -4.5597487967619E-06
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   2     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   2  1.5761147921042E+04  1.5761147182695E+04  1.5761148659397E+04
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   2    -7.3835133207467E+01 -7.3835136413574E+01  4.3422511586400E-08
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   3     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   3  1.5761147921042E+04  1.5761145083947E+04  1.5761150758261E+04
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   3    -2.8371568323564E+02 -2.8371569824219E+02  5.2892900392898E-08
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (p):   4     0     0     1    0    0   0.000000000E+00  0.000000000E+00
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (c):   4  1.5761147921042E+04  1.5761147920617E+04  1.5761147921467E+04
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk output (g):   4    -4.2512056097621E-02 -4.2511917650700E-02 -3.2566614052154E-06
tr_run.profiles/STDOUT.0000:(PID.TID 0000.0001) grdchk  summary  :  RMS of    4 ratios =  2.8018670458828E-06
(base) bkli04m056::global_oce_biogeo_bling (obsfit)> 
```

and I pushed a quick (and incomplete) fix. But it's as I said earlier. These are issues that have been fixed with PR #887 and it may be worth either waiting for that PR to be merged or at least model the obsfit files after the pkg/profiles of the corresponding branch, so that we don't have to repeat fixing the same issues again.

jm-c commented:
@averdy I could try to see how to go around this problem but it might take me some time, so I would like to have @mjlosch opinion first and see after what we could do.

mjlosch reviewed (commented):
These are just a few comments about formatting. 

I had a closer look at this variable `obs_modmask` and its toxic relationship with TAF. The variable is somewhat redundant, set to 1 if there's a valid `obs_modval` and to 0, if not. Then we pass `obs_modval` to `ACTIVE_WRITE_OBS_GLOB` as a parameter to write it to disk, but this routine also writes `obs_modmask` to disk, which is passed as a global `COMMON` variable. On top of that, this is not done for every iteration of the loops, but there are some if statements and TAF is completely confused.

 I have tried all sorts of different storing methods or `INCOMPLETE` directives, but didn't manage to have a `RECOMPUTATION Warning`-free AD-code that actually gives the correct answer (so either no warning and zero gradients, or some warnings with a correct gradient. I can also produce warnings and the wrong gradients (o:).

My suggestion is to get rid of the global variable `obs_modmask` (because it is somewhat redundant) or (if it is necessary to have in the file written to disk) at least derive it locally from the value of `obs_modval`, e.g.:
```
              obs_modmask = zeroRL
              IF ( obs_modval.GT.-9999. ) obs_modmask = oneRL
```
before it is used (near lines 305 ff), or modify it the code in a similar way, so that TAF knows what's going on. This works for me (no warnings and correct gradients).

`obsfit_globaldummy` still needs to be stored somehow, but that's unproblematic. There are a couple of other variables that are recomputed unnecessarily, but since this routine is only called once at the end of the end (or beginning of AD part), it shouldn't matter much. 
I suggest to do this with a local tape (leaving `the_main_loop.F` untouched).

@averdy I can push a version of this, if you like, but we can also first discuss, which way this should go. Maybe I have overlooked important functionality of this routine that would be jeopardised by my suggested changes.

averdy commented:
Thanks for fixing this bug @IvanaEscobar !

shoshanareich commented:
I was wondering if the second item on the "to-do list" (cost_obsfit.F can be very slow for large datasets) has been addressed yet?

averdy commented:
@jm-c I have addressed your comments 1 and 3. 
@mjlosch I've fixed some netcdf issues, so now the gradient check works with obsfitdoncoutput=.true., and mod_mask has values 0 or 1.

jm-c reviewed (approved):
LGTM

jm-c commented:
I've just pushed some minor changes (cleaning), the most significant being the last commit, as I find easier to maintain when there is less duplicated files which are not needed.
From my side, this PR looks in good shape so could move towards merging, unless @averdy you have things to add ?
And @mjlosch @jrscott do you agreee ?

averdy commented:
@jm-c Thank you! I made some changes to the manual entry, so let's wait to hear back from @jrscott .

jrscott commented:
@averdy I'll review this coming week 👍

ifenty commented:
@jrscott I can't interpret the comment log. Did you review on October 3 and, if so, can this PR move forward?
@averdy @jm-c

jm-c commented:
@ifenty and @averdy When do you need this PR to be merged into master branch ? It was pretty much ready on Sep 30, we were supposed to get some review to the docs withing the week following Oct 03, but it you prefer to have this PR merged soon we could always refine the docs in an other PR.

ifenty commented:
> @ifenty the plan is to merge this sometime next week, pending some typo fixes by @averdy and master branch merger by @jm-c

Fantastic.  There is strong interest on our side to start using the package - or at least see what we need to do to adapt it to work on our grid.
